crates/iceberg/src/spec/values.rs

```diff
@@ -996,10 +1039,552 @@ mod timestamptz {
+                            // In deserialize, the element always be `Som`e. `None` will be represented
+                            // as `Some(RawLiteral::Null)`
+                            let k_v = k_v.ok_or_else(|| invalid_err("list"))?;
+                            if let RawLiteralEnum::Record(Record {
+                                required,
+                                optional: _,
+                            }) = k_v
+                            {
+                                if required.len() != 2 {
+                                    return Err(invalid_err("list"));
+                                }
+                                let mut key = None;
+                                let mut value = None;
+                                required.into_iter().for_each(|(k, v)| {
+                                    if k == "key" {
+                                        key = Some(v);
+                                    } else if k == "value" {
+                                        value = Some(v);
+                                    }
+                                });
+                                match (key, value) {
+                                    (Some(k), Some(v)) => {
+                                        map.insert(
+                                            k.try_into(key_ty)?
+                                                .ok_or_else(|| invalid_err("list"))?,
+                                            v.try_into(value_ty)?,
+                                        );
+                                    }
+                                    _ => return Err(invalid_err("list")),
+                                }
+                            } else {
+                                return Err(invalid_err("list"));
+                            }
+                        }
+                        Ok(Some(Literal::Map(map)))
+                    }
+                    _ => Err(invalid_err("list")),
+                },
+                RawLiteralEnum::Record(Record {
+                    required,
+                    optional: _,
+                }) => match ty {
+                    Type::Struct(struct_ty) => {
+                        let iters: Vec<(i32, Option<Literal>, String)> = required
+                            .into_iter()
+                            .map(|(field_name, value)| {
+                                let field = struct_ty
+                                    .field_by_name(field_name.as_str())
+                                    .ok_or_else(|| invalid_err("record"))?;
+                                let value = value.try_into(&field.field_type)?;
+                                Ok((field.id, value, field.name.clone()))
+                            })
+                            .collect::<Result<_, Error>>()?;
+                        Ok(Some(Literal::Struct(super::Struct::from_iter(iters))))
+                    }
+                    Type::Map(map_ty) => {
```

Comment:
Is this possible? When will this happen?

Reply:
[We will serialize the map with no-string key into array of record.](https://iceberg.apache.org/spec/#avro:~:text=array%20of%20key%2Dvalue%20records) So in deserialize, it will be convert to a `RawLiteral::List` and then we convert it to `Literal::Map`. And that's why we need type info here, we need to know `RawLiteral::List` should convert to `Literal::List` or `Literal::Map`.

Reply:
I see, thanks!